metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -256,19 +271,30 @@ final class FileDecoderProvider(
 
   private def findClassesDirFileFromSource(
       sourceFile: AbsolutePath,
-      newExtension: String
-  ): Option[PathInfo] = {
+      newExtension: String,
+      position: l.Position = new l.Position(1, 1)
+  ): Option[PathInfo] =
     for {
       targetId <- buildTargets.sourceBuildTargets(sourceFile).headOption
       target <- buildTargets.scalaTarget(targetId)
       sourceRoot <- buildTargets.inverseSourceItem(sourceFile)
-      classDir = target.classDirectory.toAbsolutePath
-      oldExtension = sourceFile.extension
-      relativePath = sourceFile
-        .toRelative(sourceRoot)
-        .resolveSibling(_.stripSuffix(oldExtension) + newExtension)
-    } yield PathInfo(Some(targetId), classDir.resolve(relativePath))
-  }
+      // we use a classFinder here because file structure can differ than actual packages
+      pathToClass <- classFinder
+        .findClass(sourceFile, position)
+        .map { className =>
+          val dir = className.replace('.', '/')
+          // only .class files contains $ in name
+          val stripped =
+            if (newExtension == "tasty") dir.stripSuffix("$")
+            else dir
+          s"$stripped.$newExtension"
```

Comment:
So I think stuff still gets screwed up in here a bit. For example given a structure like:

```
❯ tree -L 4
.

├── src
│  └── main
│     └── scala
│        ├── Main.scala
│        └── Thing.scala
```

And inside of `Thing.scala` let's say you have:

```
package sanity

object Thing {
  val x = 3
}
```
The `sourceroot` ends up being `../src/main/scala`. So when it looks for the `className`, `findClass` here is actually returning `sanity`.  So when we try to put this together here we end up looking for `sanity.class`. So it's really close, but it really needs to instead look for `/sanity/Thing.class`.

Reply:
Also the same thing happens if it's in a `sanity` dir.

Reply:
Yeah, that's the same issue basically as https://github.com/scalameta/metals/issues/3136 we need to fix it in another pr I think

Reply:
Yeah, I've realized too that current implementation breaks on few scenarios :/ 
The decode file command had the same issues as show tasty, so I thought that this will help. But `ClassFinder`, in addition to sourceFile, takes also a position in the file. I thought that fallback to the `Position(1,1)` will work, but it turns out that it doesn't - that's why it's not working for your examples. For now, `ClassFinder` should be used only when a position is available.

Unfortunately, there are more issues :/ Now "show tasty" is very sensitive to the cursor position. ClassFinder for the following code will return just an empty string
```scala
package a.b
def foo = ()
>>
```
when for this one it will return the correct top-level name `a.b.<filename>$package` which then can be easily converted into the directory by replacing dots with slashes
```scala
package a.b
def foo = ()
>>
def foo2 = ()
```
For now, a simple "fallback to the filename" strategy (former behavior) should be sufficient if `ClassFinder` can't figure anything out.